File: ext/makiri/lexbor_compat/text_index.c
#include "compat.h"
#include "compat_internal.h"
#include "../core/mkr_core.h"

#include <stdint.h>
#include <stdlib.h>
#include <string.h>

/*
 * Per-document text-extraction index (lexbor_compat/text_index.c).
 *
 * Descendant-text aggregation (Node#text, XPath string-value) walks every node
 * of a subtree chasing pointers through Lexbor's 96-byte nodes - at scale this
 * is cache-bound and dominates the cost. This index removes the per-extraction
 * walk: one build pass records, in document order, a flat array of every
 * TEXT/CDATA node's byte slice, plus a pointer-keyed map from each element /
 * fragment to the contiguous [start, end) run of slices its subtree owns. A
 * later Node#text is then a hash lookup + a single pre-sized memcpy run over a
 * cache-dense slice array, touching no element nodes at all.
 *
 * Built lazily on the first text query and cached on mkr_parsed_t.text_index.
 * Invalidated (dropped) by every mutation through the same single hook as the
 * attr/element index, so a cached slice can never point at reallocated or
 * detached text storage: any structural / content / attribute change rebuilds
 * it on the next query. The slices borrow into each text node's lexbor_str_t
 * data, which the arena keeps alive (nodes are detached, never destroyed) and
 * which only a mutation can reallocate. Fail-closed: a build OOM leaves the
 * index unbuilt and the caller falls back to its own walk.
 */

typedef struct {
    const lxb_dom_node_t *node;  /* key; NULL marks an empty slot */
    uint32_t start;              /* [start, end) into slices[] (indices, not */
    uint32_t end;                /* byte offsets; bounded by the slice count, */
                                 /* which the build caps at UINT32_MAX) */
} mkr_text_range_t;

typedef struct {
    mkr_borrowed_text_t *slices; /* doc-order TEXT/CDATA slices (borrowed) */
    size_t              *prefix; /* nslices + 1; prefix[i] = bytes before slice i */
    size_t               nslices;

    mkr_text_range_t *ranges;    /* element/fragment -> slice run, open addressing */
    size_t            ranges_cap;   /* power of two, or 0 */
    size_t            ranges_count;

    lxb_dom_node_t   *root;      /* the indexed subtree root (document root element) */
} mkr_text_idx_t;

/* Explicit (heap) DFS frame - recursion is avoided so a deep tree cannot blow
 * the C stack (matches the attr/element index discipline). */
typedef struct {
    lxb_dom_node_t *node;
    lxb_dom_node_t *child;   /* next child to visit */
    size_t          range;   /* index into ranges[] for this open element */
} mkr_tix_frame_t;

/* ------------------------------------------------------------------ */
/* hashing                                                            */
/* ------------------------------------------------------------------ */

/* pow2-ceil sizing is shared: mkr_pow2_ceil (core/mkr_core.h). */

/* Bucket index for a node key (mkr_ptr_hash spreads aligned pointers). */
static size_t
mkr_tix_slot_for(const mkr_text_idx_t *t, const lxb_dom_node_t *node)
{
    return (size_t)mkr_ptr_hash(node) & (t->ranges_cap - 1);
}

/* Insert node with start (end filled on subtree close); return its slot index.
 * The table is pre-sized for exactly the element count with load < 3/4, so it
 * never rehashes mid-build and a slot is always free (linear probing). */
static size_t
mkr_tix_range_insert(mkr_text_idx_t *t, lxb_dom_node_t *node, size_t start)
{
    size_t i = mkr_tix_slot_for(t, node);
    while (t->ranges[i].node != NULL) {
        i = (i + 1) & (t->ranges_cap - 1);
    }
    t->ranges[i].node  = node;
    t->ranges[i].start = (uint32_t)start;
    t->ranges[i].end   = (uint32_t)start; /* provisional until close */
    t->ranges_count++;
    return i;
}

static const mkr_text_range_t *
mkr_tix_range_lookup(const mkr_text_idx_t *t, const lxb_dom_node_t *node)
{
    if (t->ranges_cap == 0) return NULL;
    size_t i = mkr_tix_slot_for(t, node);
    while (t->ranges[i].node != NULL) {
        if (t->ranges[i].node == node) return &t->ranges[i];
        i = (i + 1) & (t->ranges_cap - 1);
    }
    return NULL;
}

/* ------------------------------------------------------------------ */
/* build                                                              */
/* ------------------------------------------------------------------ */

static int
mkr_tix_is_text(const lxb_dom_node_t *n)
{
    return n->type == LXB_DOM_NODE_TYPE_TEXT
        || n->type == LXB_DOM_NODE_TYPE_CDATA_SECTION;
}

static int
mkr_tix_is_container(const lxb_dom_node_t *n)
{
    return n->type == LXB_DOM_NODE_TYPE_ELEMENT
        || n->type == LXB_DOM_NODE_TYPE_DOCUMENT_FRAGMENT;
}

/* Pass 1: count text slices and container (element/fragment) nodes under root
 * (inclusive), so the slice arrays and the range table are each sized once. */
static void
mkr_tix_count(lxb_dom_node_t *root, size_t *out_slices, size_t *out_containers)
{
    size_t ns = 0, nc = 0;
    for (lxb_dom_node_t *n = root; n != NULL; n = mkr_dom_preorder_next(n, root)) {
        if (mkr_tix_is_container(n)) {
            nc++;
        } else if (mkr_tix_is_text(n)) {
            const lexbor_str_t *d = &lxb_dom_interface_character_data(n)->data;
            if (d->data != NULL && d->length != 0) ns++;
        }
    }
    *out_slices = ns;
    *out_containers = nc;
}

static void
mkr_text_idx_destroy(mkr_text_idx_t *t)
{
    if (t == NULL) return;
    free(t->slices);
    free(t->prefix);
    free(t->ranges);
    free(t);
}

/* Build the index over root (the document root element). Returns NULL on OOM
 * (fail-closed: caller falls back to its own walk). */
static mkr_text_idx_t *
mkr_text_idx_build(lxb_dom_node_t *root)
{
    size_t nslices = 0, ncont = 0;
    mkr_tix_count(root, &nslices, &ncont);

    /* Slice run bounds (start/end) are stored as uint32 in the range table. A
     * document with > UINT32_MAX text slices is impossible in practice (each is
     * >= 1 byte), but guard it anyway: fail closed so the caller walks. */
    if (nslices > UINT32_MAX) return NULL;

    mkr_text_idx_t *t = mkr_callocarray(1, sizeof(*t));
    if (t == NULL) return NULL;
    t->root = root;

    if (nslices > 0) {
        t->slices = mkr_callocarray(nslices, sizeof(*t->slices));
        if (t->slices == NULL) { mkr_text_idx_destroy(t); return NULL; }
    }
    /* prefix is always allocated (>= 1 entry, prefix[0] == 0) so an empty range
     * [0,0) over a text-free subtree reads prefix[0] rather than NULL. */
    size_t pcount;
    if (!mkr_size_add(nslices, 1, &pcount)) { mkr_text_idx_destroy(t); return NULL; }
    t->prefix = mkr_callocarray(pcount, sizeof(*t->prefix));
    if (t->prefix == NULL) { mkr_text_idx_destroy(t); return NULL; }
    /* Size the range table for ncont entries at load factor < 3/4. */
    if (ncont > 0) {
        size_t want;
        if (!mkr_size_add(ncont, ncont >> 1, &want)
            || !mkr_size_add(want, 1, &want)
            || !mkr_pow2_ceil(want, &t->ranges_cap)) { mkr_text_idx_destroy(t); return NULL; }
        t->ranges = mkr_callocarray(t->ranges_cap, sizeof(*t->ranges));
        if (t->ranges == NULL) { mkr_text_idx_destroy(t); return NULL; }
    }

    /* Pass 2: explicit (heap) DFS recording each container's slice run. The
     * frame stack grows via the overflow-checked helper; depth is bounded by
     * tree depth, so it never blows the C stack. */
    mkr_tix_frame_t *st = NULL;
    size_t scap = 0, sn = 0;

    /* Seed with root (a container). */
    {
        size_t r = mkr_tix_range_insert(t, root, 0);
        if (mkr_grow_reserve((void **)&st, &scap, 1, sizeof(*st)) != MKR_OK) {
            mkr_text_idx_destroy(t); return NULL;
        }
        st[sn++] = (mkr_tix_frame_t){ root, root->first_child, r };
    }

    while (sn > 0) {
        mkr_tix_frame_t *top = &st[sn - 1];
        lxb_dom_node_t *child = top->child;
        if (child == NULL) {
            t->ranges[top->range].end = (uint32_t)t->nslices; /* close subtree */
            sn--;
            continue;
        }
        top->child = child->next; /* advance the cursor for our return */

        if (mkr_tix_is_text(child)) {
            const lexbor_str_t *d = &lxb_dom_interface_character_data(child)->data;
            if (d->data != NULL && d->length != 0) {
                size_t i = t->nslices;
                size_t total;
                if (!mkr_size_add(t->prefix[i], d->length, &total)) {
                    /* cumulative text length overflows size_t - give up the
                     * index (fail closed); the caller falls back to a walk. */
                    free(st);
                    mkr_text_idx_destroy(t);
                    return NULL;
                }
                t->slices[i].ptr = (const char *)d->data;
                t->slices[i].len = d->length;
                t->prefix[i + 1] = total;
                t->nslices = i + 1;
            }
        } else if (mkr_tix_is_container(child)) {
            size_t r = mkr_tix_range_insert(t, child, t->nslices);
            /* grow_reserve may move st; top is re-fetched at the loop head and
             * not used after this point in the iteration. */
            if (mkr_grow_reserve((void **)&st, &scap, sn + 1, sizeof(*st)) != MKR_OK) {
                free(st); mkr_text_idx_destroy(t); return NULL;
            }
            st[sn++] = (mkr_tix_frame_t){ child, child->first_child, r };
        }
        /* other kinds (comment/PI/doctype) are childless leaves: ignore. */
    }

    free(st);
    return t;
}

/* ------------------------------------------------------------------ */
/* public surface                                                     */
/* ------------------------------------------------------------------ */

void
mkr_text_index_free(void *idx)
{
    mkr_text_idx_destroy((mkr_text_idx_t *)idx);
}

void
mkr_parsed_text_index_invalidate(mkr_parsed_t *p)
{
    if (p == NULL || p->text_index == NULL) return;
    mkr_text_idx_destroy((mkr_text_idx_t *)p->text_index);
    p->text_index = NULL;
}

int
mkr_parsed_text_slices(mkr_parsed_t *p, const lxb_dom_node_t *node,
                       const mkr_borrowed_text_t **out_slices, size_t *out_n,
                       size_t *out_bytes)
{
    if (p == NULL || p->doc == NULL || node == NULL) return 0;

    mkr_text_idx_t *t = (mkr_text_idx_t *)p->text_index;
    if (t == NULL) {
        lxb_dom_node_t *root = lxb_dom_document_root((lxb_dom_document_t *)p->doc);
        if (root == NULL) return 0;
        t = mkr_text_idx_build(root);
        if (t == NULL) return 0; /* OOM: caller walks instead */
        p->text_index = t;
    }

    const mkr_text_range_t *r = mkr_tix_range_lookup(t, node);
    if (r == NULL) return 0; /* node not in the indexed tree (fragment, etc.) */

    *out_slices = t->slices + r->start;
    *out_n      = (size_t)(r->end - r->start);
    *out_bytes  = t->prefix[r->end] - t->prefix[r->start];
    return 1;
}

File: ext/makiri/rust/src/dom_adapter/text_index.rs
//! Per-document text-extraction index (dom_adapter/text_index.c).
//!
//! Descendant-text aggregation (`Node#text`, the XPath string-value) otherwise
//! walks every node of a subtree chasing pointers through Lexbor's 96-byte
//! nodes; at scale that is cache-bound and dominates the cost. This index
//! removes the per-extraction walk: one build pass records, in document order, a
//! flat array of every TEXT/CDATA node's byte slice, plus a pointer-keyed map
//! from each element or fragment to the contiguous `[start, end)` run of slices
//! its subtree owns. A later `Node#text` is a hash lookup and a single pre-sized
//! memcpy run over a cache-dense slice array, touching no element node at all.
//!
//! # The slices are raw pointers on purpose
//!
//! A borrowed `&[u8]` here would be a stronger claim than the truth. The nodes
//! belong to C, and C's mutation API writes through them outside Rust's borrow
//! checker, so a long-lived Rust reference would assert a lifetime nothing
//! enforces. What actually keeps a cached slice valid is a RUNTIME protocol:
//! every mutation goes through one invalidation hook
//! (`Parsed::invalidate_indexes`), which drops the whole index, so a
//! slice can never outlive the storage it points into. The arena also never
//! frees a node - detached, never destroyed - so only a mutation can reallocate
//! the text a slice borrows. That protocol is the safety argument; the types
//! record it rather than pretending to prove it.
//!
//! # Fail-closed
//!
//! A build that cannot allocate leaves the index unbuilt and answers 0, and the
//! caller falls back to its own walk. It must never answer a SHORTER run: a
//! truncated text is a well-formed wrong answer, which is exactly what the
//! project's fail-closed rule is about.

#![allow(unsafe_code)]
#![allow(clippy::missing_safety_doc)]

use crate::falloc::{try_vec_with_capacity, Reserve};
use crate::lexbor_abi::{self as lxb, preorder_next, LxbNode};
use crate::text::BorrowedText;
use crate::xpath::runtime_abi::cache::ptr_hash;

mod ty {
    pub use crate::dom_adapter::html::{
        TYPE_CDATA as CDATA, TYPE_ELEMENT as ELEMENT, TYPE_FRAGMENT as FRAGMENT, TYPE_TEXT as TEXT,
    };
}

/// One container's slice run. `start`/`end` are INDICES into `slices`, not byte
/// offsets, and are `u32` because the build refuses to index a document with
/// more than `u32::MAX` slices (see [`TextIndex::build`]).
#[derive(Clone, Copy)]
struct Range {
    /// The key; null marks an empty slot.
    node: *const LxbNode,
    start: u32,
    end: u32,
}

const EMPTY_RANGE: Range = Range {
    node: core::ptr::null(),
    start: 0,
    end: 0,
};

pub struct TextIndex {
    /// Document-order TEXT/CDATA slices, borrowed from the arena.
    slices: Vec<BorrowedText>,
    /// `slices.len() + 1` entries; `prefix[i]` is the byte count before slice
    /// `i`. Always non-empty, so an empty `[0, 0)` run over a text-free subtree
    /// reads `prefix[0]` rather than nothing.
    prefix: Vec<usize>,
    /// Container -> slice run, open addressing with linear probing. Empty (cap
    /// 0) when the subtree has no containers.
    ranges: Vec<Range>,
    /// A power of two, or 0.
    ranges_cap: usize,
}

impl TextIndex {
    #[inline]
    fn slot_for(&self, node: *const LxbNode) -> usize {
        (ptr_hash(node) as usize) & (self.ranges_cap - 1)
    }

    /// Insert `node` with `start`; `end` is filled when its subtree closes.
    /// Returns the slot index.
    ///
    /// The table is pre-sized for exactly the container count at load < 3/4, so
    /// it never rehashes mid-build and a free slot always exists - which is what
    /// makes this probe loop terminate.
    fn range_insert(&mut self, node: *const LxbNode, start: u32) -> usize {
        let mut i = self.slot_for(node);
        while !self.ranges[i].node.is_null() {
            i = (i + 1) & (self.ranges_cap - 1);
        }
        self.ranges[i] = Range {
            node,
            start,
            end: start,
        };
        i
    }

    fn range_lookup(&self, node: *const LxbNode) -> Option<Range> {
        if self.ranges_cap == 0 {
            return None;
        }
        let mut i = self.slot_for(node);
        while !self.ranges[i].node.is_null() {
            if self.ranges[i].node == node {
                return Some(self.ranges[i]);
            }
            i = (i + 1) & (self.ranges_cap - 1);
        }
        None
    }
}

#[inline]
unsafe fn is_container(n: *const LxbNode) -> bool {
    (*n).type_ == ty::ELEMENT || (*n).type_ == ty::FRAGMENT
}

/// The non-empty character-data payload of a TEXT/CDATA node, as `(ptr, len)`,
/// or `None`.
///
/// The single "does this node contribute a slice" test, shared by the counting
/// and filling passes so the two-pass sizing and the fill can never disagree
/// about which nodes yield one - a disagreement would size the array for a
/// different set than it fills.
#[inline]
unsafe fn text_slice(n: *const LxbNode) -> Option<(*const u8, usize)> {
    if (*n).type_ != ty::TEXT && (*n).type_ != ty::CDATA {
        return None;
    }
    let d = &(*(n as *const lxb::lxb_dom_character_data_t)).data;
    if d.data.is_null() || d.length == 0 {
        None
    } else {
        Some((d.data, d.length))
    }
}

/// Pass 1: count the text slices and the containers under `root` (inclusive),
/// so each array is sized exactly once.
unsafe fn count(root: *mut LxbNode) -> (usize, usize) {
    let (mut slices, mut containers) = (0usize, 0usize);
    let mut n = root;
    while !n.is_null() {
        if is_container(n) {
            containers += 1;
        } else if text_slice(n).is_some() {
            slices += 1;
        }
        n = preorder_next(n, root);
    }
    (slices, containers)
}

/// An explicit DFS frame. Recursion is avoided so a deep tree cannot exhaust the
/// stack - the same discipline as the attr/element index.
struct Frame {
    /// The next child to visit.
    child: *mut LxbNode,
    /// Index into `ranges` for this open container.
    range: usize,
}

impl TextIndex {
    /// Build over `root` (the document root element). `None` on OOM, which is
    /// fail-closed: the caller walks instead.
    ///
    /// # Safety
    /// `root` must be the root element of a live Lexbor document.
    pub(crate) unsafe fn build(root: *mut LxbNode) -> Option<TextIndex> {
        let (nslices, ncont) = count(root);

        /* Run bounds are u32 in the range table. A document with more than
         * u32::MAX text slices is impossible in practice (each is >= 1 byte),
         * but guard it anyway rather than truncate the index. */
        if nslices > u32::MAX as usize {
            return None;
        }

        /* Both arrays are sized EXACTLY here, so every push below lands in
         * reserved capacity and cannot allocate. That is why they use `push`
         * rather than falloc's `mkr_push`: a reserve per text node would make
         * every slice its own injection point in `rake oom` - hundreds of them
         * for one document, all testing the same branch. See clippy.toml on why
         * `push` after a successful reserve is deliberately not banned. */
        let mut t = TextIndex {
            slices: try_vec_with_capacity(nslices)?,
            prefix: try_vec_with_capacity(nslices.checked_add(1)?)?,
            ranges: Vec::new(),
            ranges_cap: 0,
        };
        t.prefix.push(0);

        if ncont > 0 {
            /* Load factor < 3/4: ncont + ncont/2 + 1, rounded up to a power of
             * two. `checked_next_power_of_two` is the fail-closed sizer - a
             * table sized below the element count would never find a free slot
             * under linear probing, and the insert loop would spin forever. */
            let want = ncont.checked_add(ncont >> 1)?.checked_add(1)?;
            let cap = want.checked_next_power_of_two()?;
            t.ranges = try_vec_with_capacity(cap)?;
            t.ranges.resize(cap, EMPTY_RANGE); /* reserved above; cannot allocate */
            t.ranges_cap = cap;
        }

        /* Pass 2: explicit DFS recording each container's slice run. The frame
         * stack is the one array whose size is not known in advance (it is
         * bounded by tree DEPTH, not node count), so it grows through falloc. */
        let mut stack: Vec<Frame> = try_vec_with_capacity(1)?;
        let r = t.range_insert(root, 0);
        stack.push(Frame {
            child: (*root).first_child,
            range: r,
        });

        while let Some(top) = stack.last_mut() {
            let child = top.child;
            if child.is_null() {
                /* Close this subtree. */
                let range = top.range;
                t.ranges[range].end = t.slices.len() as u32;
                stack.pop();
                continue;
            }
            top.child = (*child).next; /* advance the cursor for our return */

            if let Some((ptr, len)) = text_slice(child) {
                /* Compared against the COUNT, not against `capacity()`: a Vec
                 * may reserve more than asked, so capacity would not catch a
                 * count/fill disagreement. One more slice than pass 1 saw means
                 * the tree changed under us - fail closed rather than push past
                 * the reservation. */
                if t.slices.len() == nslices {
                    return None;
                }
                /* The running total is checked rather than wrapped: a wrapped
                 * prefix would make a later `prefix[end] - prefix[start]`
                 * smaller than the bytes actually present, which is a short read
                 * into a pre-sized String. */
                let total = t.prefix[t.slices.len()].checked_add(len)?;
                t.slices.push(unsafe {
                    BorrowedText::from_raw_parts(ptr as *const core::ffi::c_char, len)
                });
                t.prefix.push(total);
            } else if is_container(child) {
                let start = t.slices.len() as u32;
                let r = t.range_insert(child, start);
                /* Reserve only when the stack is actually full. `mkr_push`
                 * consults the injection counter on EVERY call, so pushing
                 * unconditionally made each of a document's containers its own
                 * injection point - 178 for one `rake oom` scenario where the C
                 * had 16, all re-testing one branch. The C grows the same way,
                 * through grow_reserve, and only when it must. */
                if stack.len() == stack.capacity() {
                    let want = crate::falloc::grow_capacity(
                        stack.capacity(),
                        stack.len() + 1,
                        core::mem::size_of::<Frame>(),
                    )?;
                    stack.mkr_reserve_exact(want - stack.len()).ok()?;
                }
                stack.push(Frame {
                    child: (*child).first_child,
                    range: r,
                });
            }
            /* Other kinds (comment / PI / doctype) are childless leaves. */
        }

        Some(t)
    }
}

/* ------------------------------------------------------------------ *
 * public surface                                                     *
 * ------------------------------------------------------------------ */

impl TextIndex {
    /// The document-order run of text slices `node`'s subtree owns, with its
    /// byte total; None for a node outside the indexed tree. Never a shorter
    /// run than the truth.
    pub fn slices_of(&self, node: *const LxbNode) -> Option<(&[BorrowedText], usize)> {
        let r = self.range_lookup(node)?;
        let (start, end) = (r.start as usize, r.end as usize);
        Some((
            &self.slices[start..end],
            self.prefix[end] - self.prefix[start],
        ))
    }
}
